src/wmtk/utils/polygon_mesh_topology_initialization.cpp

```diff
@@ -0,0 +1,409 @@
+VectorXl build_inverse(Eigen::Ref<const VectorXl> f, long size)
+{
+    VectorXl g = VectorXl::Constant(size, 1, -1);
+    for (long i = 0; i < f.size(); ++i) {
+        // Ignore negative (invalid/boundary) indices
+        if (f(i) < 0) {
+            continue;
+        }
+        g(f(i)) = i;
+    }
+
+    return g;
+}
+
+// Build a right inverse g:{0,...,m-1}->{0,...,n-1} for a function f:{0,...,n-1}->{-l,...,0,...,m-1}
+// so that f(g(i)) = i.
+//
+// Negative indices in f are ignored, and g(i) is set to -1 if i is not in the image of f.
+VectorXl build_right_inverse(Eigen::Ref<const VectorXl> f)
+{
+    long n = f.maxCoeff() + 1;
+    return build_inverse(f, n);
+}
+
+// Build an inverse g for a function f. Requires f to be invertible.
+VectorXl build_inverse(Eigen::Ref<const VectorXl> f)
+{
+    long n = f.size();
+    return build_inverse(f, n);
+}
+
+// Permute an index map m on the left by a permutation g (i.e., i -> g(m(i))). Ignore negative
+// indices
+VectorXl reindex_map_image(Eigen::Ref<const VectorXl> map, Eigen::Ref<const VectorXl> perm)
+{
+    long domain_size = map.size();
+    VectorXl permuted_map = map;
+    for (long i = 0; i < domain_size; ++i) {
+        // Leave negative indices unchanged
+        if (map(i) >= 0) {
+            permuted_map(i) = perm(map(i));
+        }
+    }
+
+    return permuted_map;
+}
+
+// Permute an index map m on the right by a permutation g (i.e., i -> m(g^{-1}(i))))
+VectorXl reindex_map_domain(Eigen::Ref<const VectorXl> map, Eigen::Ref<const VectorXl> perm)
+{
+    long domain_size = map.size();
+    VectorXl permuted_map(domain_size);
+    for (long i = 0; i < domain_size; ++i) {
+        permuted_map(perm(i)) = map(i);
+    }
+
+    return permuted_map;
+}
+
+// Conjugate an index map m on the by a permutation g (i.e., i -> g(m(g^{-1}(i)))))
+VectorXl conjugate_permutation_map(Eigen::Ref<const VectorXl> map, Eigen::Ref<const VectorXl> perm)
+{
+    return reindex_map_domain(reindex_map_image(map, perm), perm);
+}
+
+// Build the opposite halfedge map from halfedge to tail (from) and head (to) vertex maps,
+// gluing together halfedges with common endpoints
+VectorXl build_opp(Eigen::Ref<const VectorXl> from, Eigen::Ref<const VectorXl> to)
+{
+    assert(to.size() == from.size());
+    long n_he = from.size();
+    long n_v = from.maxCoeff() + 1;
+
+    // Shift indices by 1 to distinguish from 0 entries in the matrix
+    std::vector<long> he_index = range(1, n_he + 1);
+
+    // Create the adjacency matrix (from, to) -> halfedge index+1
+    Eigen::SparseMatrix<long> vv2he(n_v, n_v);
+    typedef Eigen::Triplet<long> Trip;
+    std::vector<Trip> trips(n_he);
+    for (long he = 0; he < n_he; ++he) {
+        trips[he] = Trip(from(he), to(he), he_index[he]);
+    }
+    vv2he.setFromTriplets(trips.begin(), trips.end());
+
+    // Create opp array
+    // TODO Check validity and manifold
+    VectorXl opp(n_he);
+    for (long he = 0; he < n_he; ++he) {
```

Comment:
in retrospect ... why do we construct an actual sparse matrix? the values in trips are all unique - might as well read off from `from(he),to(he), he_index[he]` directly?